### ISSUE TITLE
Product hasn't been defined before - anyone had this issue previously?

### ISSUE THREAD

**Issue by dazmahal** [xBimTeam/XbimWebUI#53]
Hi Folks

Receiving this error when trying to Load Model from latest Master Branch Web UI
Any ideas on resolution, should I exclude the shapes from the model on conversion?
Thanks

![image](https://cloud.githubusercontent.com/assets/6757134/11446289/87fda4be-952c-11e5-8699-fc1664f96bdb.png)

**Comment by mbacon-edocs** [comment]
See if your model loads properly if you change that line to be:

`if (typeof (map) === "undefined") continue;//throw "Product hasn't been defined before.";`

This isn't a permanent fix, but will go to towards confirming that this is the same issue that I had last week. On discovery of this issue there could be a big pull request coming from me at some point.

**Comment by martin1cerny** [comment]
Hi Darren,

This issue is caused by the new Geometry Engine which apparently writes out openings geometry but not their products. It should either write out both or none (it is not necessary for most scenarios as openings are cut out from walls and other elements already). This is an inconsistent wexBIM data file. I’ll check with Steve to find out where this had happened. Until than you could try to reference one version older geometry engine which worked fine.

Will let you know once it is resolved.

Regards
Martin

**Comment by mbacon-edocs** [comment]
Hi @martin1cerny, just to let you know that whilst that could be the issue here, for me it wasn't. Instead it was down to a problem with the code not being defensive enough and being affected by another 3rd party library. The problem exists in quite a few places and is something that I will look at creating a pull request for at some point, but there's a lot to change.

**Comment by martin1cerny** [comment]
Hi @mbacon-edocs , 

this was exactly the case of inconsisten wexBIM file. I've put in a temporal fix until geometry engine is not fixed (already published on MyGet in master and develop branch).

Could you be more specific about the nature of problem you describe here? What was that 3rd party library that caused the problem? Looking forward for your improvements.

Martin

**Comment by mbacon-edocs** [comment]
Hi @martin1cerny,

The problem was with MooTools which was included in a 360 panorama suite. However, the problem could potentially arise with a multitude of libraries and custom user code. The main reason for this is the use of the `for..in` loop which ends up iterating over unwanted items. See [this stackoverflow question](http://stackoverflow.com/questions/500504/why-is-using-for-in-with-array-iteration-such-a-bad-idea) - the top answer, it applies to arrays and objects.

There are two main alternatives:
```javascript
//Opt 1
for(var key in collection)
{
    if(!collection.hasOwnProperty(key)) continue;
    console.log(key, collection[key]);
}
//Opt 2 (IE9+)
var keys = Object.getOwnPropertyNames(collection);
keys.forEach(function(key) {
    console.log(key, collection[key]);
});
```

**Comment by martin1cerny** [comment]
This was fixed by pull request #85 .
